File: frontend/lib/selectividadMargen.ts
// =====================================================================
// El margen de selectividad: ΔΔG, y UNA sola escala
// =====================================================================
//
// Espejo exacto de `backend/services/docking/selectividad_margen.py`. Hay una
// prueba que compara los umbrales de los dos archivos, porque lo que se está
// arreglando es justamente que existían por duplicado y no coincidían:
//
//     backend  (selectivity_verdict)   > 10   > 3    > 1.5  > 1.0
//     frontend (ProSelectivityPanel)   > 1.8  > 1.2  > 0.9
//
// Un cociente de 2.0 se guardaba en la base y en el dossier como «MODERADAMENTE
// SELECTIVO» mientras la pantalla decía «ALTAMENTE SELECTIVO» para la MISMA
// corrida. Quien exportaba el dossier leía un veredicto y quien miraba el panel
// leía otro.
//
// ─── Y el cociente en sí tampoco servía ─────────────────────────────────
//
// `ratio = ΔG_on / ΔG_off` divide dos energías libres. Eso no es una razón de
// selectividad: como ΔG = −RT·ln K_d, el cociente es ln K_on / ln K_off, el
// logaritmo de una constante de disociación en la base de la otra. No describe
// nada.
//
//     ΔG_on = −10.0, ΔG_off = −5.0   ->  cociente 2.00   ->  ΔΔG = 5.0 kcal/mol
//     ΔG_on =  −6.0, ΔG_off = −3.0   ->  cociente 2.00   ->  ΔΔG = 3.0 kcal/mol
//
// Mismo número, mismo veredicto, y márgenes reales que se diferencian por un
// factor de treinta.
//
// La magnitud correcta es la diferencia: ΔΔG = ΔG_off − ΔG_on, en kcal/mol.

/** ln(10)·R·T a 298.15 K, en kcal/mol. Convierte ΔΔG en órdenes de magnitud. */
export const KCAL_POR_DECADA = 1.3633;

/** Dispersión típica del score de Vina frente a afinidad medida (kcal/mol). */
export const INCERTIDUMBRE_VINA_KCAL = 2.0;

/** Cuántas veces más fuerte es la unión a la diana principal. */
export function factorDeSelectividad(deltaDeltaG: number): number {
  return 10 ** (deltaDeltaG / KCAL_POR_DECADA);
}

/** ΔΔG = ΔG_off − ΔG_on. Positivo = la diana principal une más fuerte. */
export function margenDeSelectividad(
  afinidadOn: number | null | undefined,
  afinidadPeorOff: number | null | undefined,
): number | null {
  if (afinidadOn == null || afinidadPeorOff == null) return null;
  return Number((afinidadPeorOff - afinidadOn).toFixed(2));
}

export const VEREDICTO_SIN_DATOS = "Sin datos suficientes";
export const VEREDICTO_INVERTIDO =
  "MARGEN NEGATIVO — la molécula une más fuerte a una anti-diana que a su diana";

/**
 * Los cortes están en órdenes de magnitud redondos, no elegidos a ojo:
 * 2.73 kcal/mol son 100x, 1.36 son 10x, y 0 es «indistinguible».
 *
 * Ninguna etiqueta dice «seguro». Un margen in silico con ±2 kcal/mol de
 * incertidumbre no autoriza esa palabra, y menos en el panel de anti-dianas.
 */
export const ESCALA_DE_MARGEN: ReadonlyArray<readonly [number, string]> = [
  [2.73, "MARGEN AMPLIO — la diana principal une ~100x más fuerte que la peor anti-diana"],
  [1.36, "MARGEN MODERADO — unas 10x a favor de la diana principal"],
  [0.41, "MARGEN ESTRECHO — menos de 10x; el orden puede invertirse dentro del error del método"],
  [0.0, "SIN MARGEN — afinidades indistinguibles entre diana y anti-diana"],
];

export function veredictoDeMargen(deltaDeltaG: number | null | undefined): string {
  if (deltaDeltaG == null) return VEREDICTO_SIN_DATOS;
  if (deltaDeltaG < 0) return VEREDICTO_INVERTIDO;
  for (const [minimo, veredicto] of ESCALA_DE_MARGEN) {
    if (deltaDeltaG >= minimo) return veredicto;
  }
  return VEREDICTO_SIN_DATOS;
}

/** El color sale de la MISMA escala, no de umbrales propios. */
export function colorDeMargen(deltaDeltaG: number | null | undefined): string {
  if (deltaDeltaG == null) return "#64748b";
  if (deltaDeltaG < 0) return "#ef4444";
  if (deltaDeltaG >= 2.73) return "#10b981";
  if (deltaDeltaG >= 1.36) return "#34d399";
  if (deltaDeltaG >= 0.41) return "#f59e0b";
  return "#ef4444";
}

/** El factor implicado, como intervalo. Un punto sería pseudoprecisión. */
export function bandaDeFactor(deltaDeltaG: number): { min: number; max: number } {
  return {
    min: factorDeSelectividad(deltaDeltaG - INCERTIDUMBRE_VINA_KCAL),
    max: factorDeSelectividad(deltaDeltaG + INCERTIDUMBRE_VINA_KCAL),
  };
}

/** «120x», «1.4x», «0.3x» — sin decimales cuando no los sostiene. */
export function formatearFactor(factor: number): string {
  if (factor >= 100) return `${Math.round(factor).toLocaleString("es")}x`;
  if (factor >= 10) return `${factor.toFixed(0)}x`;
  return `${factor.toFixed(1)}x`;
}

File: backend/services/docking/selectividad_margen.py
"""
El margen de selectividad, en la magnitud que sí tiene sentido físico.

# El fallo

    selectivity_ratio = round(on_target_affinity / worst_off_affinity, 2)

Las dos cantidades son energías libres de unión en kcal/mol. Dividirlas no
produce una razón de selectividad: produce un número adimensional sin
significado termodinámico. Como ΔG = −RT·ln K_d, ese cociente es
ln K_on / ln K_off, es decir el logaritmo de una constante de disociación en
la base de la otra. No es una cantidad que describa nada.

Lo que se ve en la práctica, con la escala de verdictos que había:

    ΔG_on = −10.0, ΔG_off = −5.0   ->  cociente 2.00   ->  ΔΔG = 5.0 kcal/mol
    ΔG_on =  −6.0, ΔG_off = −3.0   ->  cociente 2.00   ->  ΔΔG = 3.0 kcal/mol

Mismo cociente, mismo veredicto — y márgenes reales que se diferencian por un
factor de treinta. Y al revés: dos moléculas con el MISMO ΔΔG reciben cocientes
distintos según lo fuerte que sea la unión, así que el número ni siquiera
ordena de forma consistente.

Además el cociente se rompe justo en el mejor caso posible: si la molécula no
se une en absoluto a la anti-diana (ΔG_off ≥ 0), no hay división que hacer y el
resultado salía `null` —«sin datos suficientes»— para la situación más segura
que puede darse.

# La magnitud correcta

    ΔΔG = ΔG_off − ΔG_on          [kcal/mol]

Positivo = la diana principal une más fuerte que la anti-diana. Es una
diferencia de energías, se mide en las mismas unidades que sus términos, y es
lo que la literatura de selectividad reporta.

De ahí sale el factor de selectividad, que es lo que un químico médico quiere
leer:

    factor = 10^(ΔΔG / (ln(10)·R·T))   con ln(10)·R·T = 1.3633 kcal/mol a 298 K

# Por qué el factor va como banda y no como número

El score de Vina no es una energía libre medida: es una función empírica de
puntuación con una dispersión de ~2 kcal/mol frente a afinidad experimental.
ΔΔG hereda esa dispersión —parte se cancela al comparar dos acoplamientos del
mismo ligando, pero no toda: son receptores y cajas distintas—, y en escala
exponencial 2 kcal/mol son más de un orden de magnitud.

Dar «4.680×» sería el mismo error de pseudoprecisión que ya se quitó de la
columna de Ki. Se reporta ΔΔG en kcal/mol como cifra principal, y el factor
como intervalo.

# Una sola escala

Los umbrales del veredicto vivían por duplicado y NO COINCIDÍAN:

    backend  (selectivity_verdict)   > 10   > 3    > 1.5  > 1.0
    frontend (ProSelectivityPanel)   > 1.8  > 1.2  > 0.9

Un cociente de 2.0 se guardaba en la base y en el dossier como «MODERADAMENTE
SELECTIVO» mientras la pantalla decía «ALTAMENTE SELECTIVO» para la misma
corrida. Aquí hay una sola escala, en ΔΔG, y el frontend la reimplementa contra
esta tabla con una prueba que compara ambas.
"""

from __future__ import annotations

from dataclasses import dataclass

#: ln(10)·R·T a 298.15 K, en kcal/mol. Convierte ΔΔG en órdenes de magnitud.
KCAL_POR_DECADA: float = 1.3633

#: Dispersión típica del score de Vina frente a afinidad medida (kcal/mol).
#: La misma constante que usa el panel para la banda de concentración.
INCERTIDUMBRE_VINA_KCAL: float = 2.0


@dataclass(frozen=True)
class MargenDeSelectividad:
    """Lo que la corrida sostiene sobre el margen frente a las anti-dianas."""

    #: ΔG_off − ΔG_on, en kcal/mol. Positivo = la diana principal une más fuerte.
    delta_delta_g: float
    #: Anti-diana que define el margen (la que une más fuerte de todas).
    anti_diana: str | None
    #: Factor de selectividad implicado, y su intervalo por la dispersión de Vina.
    factor: float
    factor_min: float
    factor_max: float


def margen_de_selectividad(
    afinidad_on: float | None,
    afinidad_peor_off: float | None,
    anti_diana: str | None = None,
) -> MargenDeSelectividad | None:
    """ΔΔG y el factor que implica. `None` si falta alguna de las dos afinidades.

    A diferencia del cociente anterior, `afinidad_peor_off >= 0` NO es un caso
    degenerado: significa que la molécula no se une a la anti-diana, que es el
    mejor resultado posible, y produce un ΔΔG grande y positivo.
    """
    if afinidad_on is None or afinidad_peor_off is None:
        return None

    ddg = float(afinidad_peor_off) - float(afinidad_on)
    return MargenDeSelectividad(
        delta_delta_g=round(ddg, 2),
        anti_diana=anti_diana,
        factor=factor_de_selectividad(ddg),
        factor_min=factor_de_selectividad(ddg - INCERTIDUMBRE_VINA_KCAL),
        factor_max=factor_de_selectividad(ddg + INCERTIDUMBRE_VINA_KCAL),
    )


def factor_de_selectividad(delta_delta_g: float) -> float:
    """Cuántas veces más fuerte es la unión a la diana principal."""
    return float(10.0 ** (delta_delta_g / KCAL_POR_DECADA))


# ── La escala, en un solo sitio ────────────────────────────────────────────
#
# Los cortes están en órdenes de magnitud redondos, no en números elegidos a
# ojo: 2.73 kcal/mol son 100x, 1.36 son 10x, y 0 es «indistinguible».
#
# Ninguna etiqueta dice «seguro». Un margen in silico de dos órdenes con una
# incertidumbre de ±2 kcal/mol no autoriza esa palabra, y el panel de
# anti-dianas es justo donde no se puede insinuar.

#: (ΔΔG mínimo, veredicto). En orden descendente; el primero que se cumple gana.
ESCALA_DE_MARGEN: tuple[tuple[float, str], ...] = (
    (2.73, "MARGEN AMPLIO — la diana principal une ~100x más fuerte que la peor anti-diana"),
    (1.36, "MARGEN MODERADO — unas 10x a favor de la diana principal"),
    (0.41, "MARGEN ESTRECHO — menos de 10x; el orden puede invertirse dentro del error del método"),
    (0.00, "SIN MARGEN — afinidades indistinguibles entre diana y anti-diana"),
)

VEREDICTO_INVERTIDO = "MARGEN NEGATIVO — la molécula une más fuerte a una anti-diana que a su diana"
VEREDICTO_SIN_DATOS = "Sin datos suficientes"


def veredicto_de_margen(delta_delta_g: float | None) -> str:
    """El texto canónico. Sin API, sin base de datos, sin subprocesos."""
    if delta_delta_g is None:
        return VEREDICTO_SIN_DATOS
    if delta_delta_g < 0:
        return VEREDICTO_INVERTIDO
    for minimo, veredicto in ESCALA_DE_MARGEN:
        if delta_delta_g >= minimo:
            return veredicto
    return VEREDICTO_SIN_DATOS
